core/libp2p/protocol_muxer/multiselect/multiselect.cpp

```diff
@@ -5,269 +5,257 @@
+  void Multiselect::handleNaMsg(
+      const std::shared_ptr<ConnectionState> &connection_state) const {
     // if we receive na message, just send an ls to understand, which protocols
     // the other side supports
-    OUTCOME_TRY(connection->write(MessageManager::lsMsg()));
-    return Status::LS_SENT;
+    MessageWriter::sendLsMsg(connection_state);
+  }
+
+  void Multiselect::onProtocolAfterOpeningOrLs(
+      std::shared_ptr<ConnectionState> connection_state,
+      const peer::Protocol &protocol) {
+    // the other side wants to communicate over that protocol; if it's available
+    // on our side, round is finished
+    auto protocols_to_search = connection_state->protocols;
+    if (protocols_to_search->empty()) {
+      return negotiationRoundFailed(connection_state,
+                                    MultiselectError::INTERNAL_ERROR);
+    }
+    if (std::find(protocols_to_search->begin(), protocols_to_search->end(),
+                  protocol)
+        != protocols_to_search->end()) {
+      return MessageWriter::sendProtocolAck(std::move(connection_state),
+                                            protocol);
+    }
+
+    // if the protocol is not available, send na
+    MessageWriter::sendNaMsg(connection_state);
+  }
+
+  void Multiselect::onProtocolsAfterLs(
+      const std::shared_ptr<ConnectionState> &connection_state,
+      gsl::span<const peer::Protocol> received_protocols) {
+    // TODO(akvinikym) [PRE-143] 30.04.19: store the protocols to the
+    // protocol repository
+
+    // if any of the received protocols is supported by our side, choose it;
+    // fail otherwise
+    auto protocols_to_search = connection_state->protocols;
+    for (const auto &proto : *protocols_to_search) {
+      // as size of vectors should be around 10 or less, we can use O(n*n)
+      // approach
+      if (std::find(received_protocols.begin(), received_protocols.end(), proto)
+          != received_protocols.end()) {
+        // the protocol is found
+        return MessageWriter::sendProtocolMsg(proto, connection_state);
+      }
+    }
+
+    negotiationRoundFailed(connection_state,
+                           MultiselectError::NEGOTIATION_FAILED);
+  }
+
+  void Multiselect::onUnexpectedRequestResponse(
+      const std::shared_ptr<ConnectionState> &connection_state) const {
+    log_->info("got a unexpected request-response combination - sending 'ls'");
+    MessageWriter::sendLsMsg(connection_state);
   }
 
-  outcome::result<Multiselect::Status> Multiselect::onUnexpectedRequestResponse(
-      const std::shared_ptr<basic::ReadWriteCloser> &connection) const {
-    log_->info("got a unexpected request-response combination");
-    return Status::NEGOTIATION_FAIL;
+  void Multiselect::onGarbagedStreamStatus(
+      const std::shared_ptr<ConnectionState> &connection_state) const {
+    log_->critical("there is some garbage in stream state status");
+    MessageWriter::sendLsMsg(connection_state);
```

Comment:
```suggestion
    clearResources(connection_state);
    negotiationRoundFailed(connection_state, INTERNAL_ERROR);
```
You should not try to continue negotiation in case of such a bug

Reply:
Agree. Clear resources is done in negotiationRoundFailed